dvc/remote/base.py

```diff
@@ -541,22 +573,36 @@ def walk_files(self, path_info):
     def protect(path_info):
         pass
 
-    def save(self, path_info, checksum_info, save_link=True):
+    def save(self, path_info, checksum_info, save_link=True, tree=None):
         if path_info.scheme != self.scheme:
             raise RemoteActionNotImplemented(
                 f"save {path_info.scheme} -> {self.scheme}", self.scheme,
             )
 
-        checksum = checksum_info[self.PARAM_CHECKSUM]
-        self._save(path_info, checksum, save_link)
+        if tree:
+            # save checksum will be computed during tree walk
+            checksum = None
+        else:
+            checksum = checksum_info[self.PARAM_CHECKSUM]
+        self._save(path_info, checksum, save_link, tree)
+
+    def _save(self, path_info, checksum, save_link=True, tree=None):
+        if tree:
+            logger.debug("Saving tree path '%s' to cache.", path_info)
+        else:
+            to_info = self.checksum_to_path_info(checksum)
+            logger.debug("Saving '%s' to '%s'.", path_info, to_info)
 
-    def _save(self, path_info, checksum, save_link=True):
-        to_info = self.checksum_to_path_info(checksum)
-        logger.debug("Saving '%s' to '%s'.", path_info, to_info)
-        if self.isdir(path_info):
-            self._save_dir(path_info, checksum, save_link)
+        if tree:
+            isdir = tree.isdir
+            save_link = False
+        else:
+            isdir = self.isdir
```

Comment:
I like how RemoteTree(or something like that) becomes more and more apparent :smile:

And that the state should probably belong to the tree. So here it would be a noop for git tree.

Not saying we should implement all of that right now, just noticing the things that we've discussed earlier :slightly_smiling_face: